src/main/java/htsjdk/variant/vcf/VCFHeaderMerger.java

```diff
@@ -0,0 +1,286 @@
+ * Some headers sets cannot be merged, and will result in an exception being thrown:
+ * <ul>
+ * <li> Headers must have a version that is at least VCF v4.2. Headers from older versions may not be merged (note
+ * that older headers that are read from input files are automatically "converted" to VCF v4.2 by VCFCodec. See
+ * {@link AbstractVCFCodec#setVCFHeader(VCFHeader).}</li>
+ * <li> any header that contains a header line that doesn't conform to the resulting (highest )version of any
+ * header in the merge list </li>
+ * <li> any header that has a sequence dictionary that is incompatible with any other merged header's
+ * sequence dictionary. All headers must either share a common sequence dictionary, or have a sequence dictionary
+ * that is a subset of the common sequence dictionary that is taken from the remaining headers. </li>
+ * </ul>
+ */
+public class VCFHeaderMerger {
+
+    /**
+     * Merge all header lines in a set of headers into a single set of header lines. The resulting set includes
+     * all unique lines that appeared in any header; duplicates of lines are excluded from the result set. Equivalent
+     * header lines are reduced to a single representative header line. The resulting set contains a ##fileformat
+     * version line for the newest version seen in any of the headers provided in the input header collection,
+     * and all lines in the merged set are compatible with that version.
+     *
+     * @param headers the headers to merge
+     * @param emitWarnings true if warnings should be emitted
+     * @return a set of merged VCFHeaderLines
+     * @throws TribbleException if any header has a version < VCFv4.2, or if any header line in any
+     * input header is not compatible the newest version selected from amongst all headers provided, or if any
+     * header has a sequence dictionary that is incompatible with any other header's sequence dictionary
+     */
+    public static Set<VCFHeaderLine> getMergedHeaderLines(final Collection<VCFHeader> headers, final boolean emitWarnings) {
+        ValidationUtils.nonNull(headers, "headers");
+        ValidationUtils.validateArg(!headers.isEmpty(), "headers collection must be non empty");
+
+        // use a VCFMetaDataLines object to accumulate header lines
+        final VCFMetaDataLines mergedMetaData = new VCFMetaDataLines();
+        final HeaderMergeConflictWarnings conflictWarner = new HeaderMergeConflictWarnings(emitWarnings);
+
+        final VCFHeaderVersion newestVersion = getNewestHeaderVersion(headers);
+        final SAMSequenceDictionary commonSequenceDictionary = getCommonSequenceDictionaryOrThrow(headers, conflictWarner);
+
+        for (final VCFHeader sourceHeader : headers) {
+            for (final VCFHeaderLine line : sourceHeader.getMetaDataInSortedOrder()) {
+                final String key = line.getKey();
+                if (VCFHeaderVersion.isFormatString(key) || key.equals(VCFHeader.CONTIG_KEY)) {
```

Comment:
Removing the redundant VCFHeaderVersion from VCFHeader makes sense now I think - we'll just need to query VCFMetaDataLines before and after every change, but thats cheap.

Changing VCFMetaData lines to not model the version as a line in the list will result in awkward semantics and implementation for things like `removeMetaDataLine`, `findEquivalentHeaderLine`, `getOtherHeaderLines`, but I don't have a strong opinion on that - as long as we keep the awkward public header API, there will be awkward semantics one way or another. I'll make the change and we can see which way we think is better. 

None of that will address the external consumers though, such as `VCFWriter`, which is still going to have to strip the file format line manually, since we need to hand them out at the public boundary to be compatible with past behavior. The right long term fix is to have a real codec that encodes as well as decodes, such as we have in the CZI package.